codec/src/main/java/io/netty/handler/codec/FlowControlHandler.java

```diff
@@ -0,0 +1,245 @@
+ * by a {@code LastHttpContent} event.
+ *
+ * <pre>
+ * {@link ChannelPipeline} pipeline = ...;
+ *
+ * pipeline.addLast(new HttpServerCodec());
+ * pipeline.addLast(new {@link FlowControlHandler}());
+ *
+ * pipeline.addLast(new MyExampleHandler());
+ *
+ * class MyExampleHandler extends {@link ChannelInboundHandlerAdapter} {
+ *   @Override
+ *   public void channelRead({@link ChannelHandlerContext} ctx, Object msg) {
+ *     if (msg instanceof HttpRequest) {
+ *       ctx.channel().config().setAutoRead(false);
+ *
+ *       // The FlowControlHandler will hold any subsequent events that
+ *       // were emitted by HttpObjectDecoder until auto reading is turned
+ *       // back on or Channel#read() is being called.
+ *     }
+ *   }
+ * }
+ * </pre>
+ *
+ * @see ChannelConfig#setAutoRead(boolean)
+ */
+public class FlowControlHandler extends ChannelDuplexHandler {
+    private static final InternalLogger logger = InternalLoggerFactory.getInstance(FlowControlHandler.class);
+
+    private final boolean releaseMessages;
+
+    private RecyclableArrayDeque queue;
+
+    private ChannelConfig config;
+
+    private int minConsume;
+
+    public FlowControlHandler() {
+        this(true);
+    }
+
+    public FlowControlHandler(boolean releaseMessages) {
+        this.releaseMessages = releaseMessages;
+    }
+
+    /**
+     * Returns a copy of the underlying {@link Queue}. This method exists for
+     * testing, debugging and inspection purposes and it is not Thread safe!
+     */
+    Queue<Object> queue() {
+        RecyclableArrayDeque queue = this.queue;
+
+        if (queue == null) {
+            return new ArrayDeque<Object>(0);
+        }
+
+        return new ArrayDeque<Object>(queue);
+    }
+
+    /**
+     * Releases all messages and destroys the {@link Queue}.
+     */
+    private void destroy() {
+        if (queue != null) {
+
+            if (!queue.isEmpty()) {
+                logger.trace("Non-empty queue: {}", queue);
+
+                if (releaseMessages) {
+                    Object msg;
+                    while ((msg = queue.poll()) != null) {
+                        ReferenceCountUtil.safeRelease(msg);
+                    }
+                }
+            }
+
+            queue.recycle();
+            queue = null;
+        }
+    }
+
+    @Override
+    public void handlerAdded(ChannelHandlerContext ctx) throws Exception {
+        Channel channel = ctx.channel();
+        config = channel.config();
+    }
+
+    @Override
+    public void channelInactive(ChannelHandlerContext ctx) throws Exception {
+        destroy();
+        ctx.fireChannelInactive();
+    }
+
+    @Override
+    public void read(ChannelHandlerContext ctx) throws Exception {
+        if (dequeue(ctx, 1) == 0) {
+            // It seems the queue was left in an empty state. We
+            // need to read() more messages from upstream.
+
+            minConsume = 1;
```

Comment:
if you like `minConsume` in the `dequeue` I think that can still stay ... I was just curious about the member variable.

```java
int minConsume = this.shouldConsume ? 1 : 0;
this.shouldConsume = false;
dequeue(..., minConsume);
```

also if there it is useful for `this.minConsume` to not be `0` or `1` then maybe just leave it and expose this somehow so I can follow your train of thought?